docs/concepts/native-interop.rst

```diff
@@ -0,0 +1,260 @@
+Native interoperability
+=======================
+
+In this document we will dive a little bit deeper into all three ways of doing 
+"native interoperability" that are available on the .NET platform. 
+
+To refresh your memory, the two main ways of doing native interoperability are:
+
+* Platform Invoke (P/Invoke)
+* COM interoperability
+
+You might be wondering why would you ever want or need to do something like this. 
+After all, .NET implementations usually come with very rich class libraries and 
+with `NuGet <http://www.nuget.org/>`_ package manager you have access to even 
+more managed APIs. So, really, why mess with unmanaged code at all?
+
+There are a couple of reasons:
+
+* Operating Systems come with a large volume of APIs that are 
+  not present in the managed class libraries. A prime example for this would be 
+  access to hardware or operating system management functions.
+* Communicating with components that are written in unmanaged languages usually 
+  means using P/Invoke. This covers components written in other managed languages,
+  such as Java for instance.
+* On Windows, most of the software that gets installed, such as Microsoft Office
+  suite, registers COM components that represent their programs and allow developers 
+  to automate them or use them. This also requires native interoperability.
+
+There are certainly more situations where it is apt to use native code, especially 
+those that are specific for a given developer's scenario. 
+
+Platform Invoke (P/Invoke)
+--------------------------
+P/Invoke is a technology that allows you to access structs, callbacks and functions
+in unmanaged libraries from your managed code. Most of the P/Invoke API is contained 
+in two namespaces: ``System`` and ``System.Runtime.InteropServices``. Using these 
+two namespaces will allow you access to the attributes
+
+Let's start from the most common example, and that is reusing unmanaged functions 
+in your managed code. Let's show a message box from a command-line application:
+
+.. code-block:: c#
+    :linenos:
+    
+    using System.Runtime.InteropServices;
+
+    public class Program {
```

Comment:
IMO it would be nice to follow same guildelines in samples https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md